File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.cpp
class Solution {
public:
    int dfs(vector<vector<int>>& tree, 
            int node, int nodeCost, vector<bool>& visited, vector<bool>& hasApple) {
	    if (visited[node]) {
		    return 0;
	    }
	    visited[node] = true;
        int childrenCost = 0;
        for (auto neigh: tree[node]) {
            childrenCost += dfs(tree, neigh, 2, visited, hasApple);
        }
        if (childrenCost == 0 && !hasApple[node]) {
            return 0;
        }
        return (childrenCost + nodeCost);
    }

    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        vector<vector<int>> tree(n);
        for (auto it : edges) {
            tree[it[0]].push_back(it[1]);
            tree[it[1]].push_back(it[0]);
        }
        vector<bool> visited(n, false);
        return dfs(tree, 0, 0, visited, hasApple);
    }
};

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.java
class Solution {
    public boolean[] visited;
    
    public int minTime(int n, int[][] edges, List<Boolean> hasApple) {
        List<List<Integer>> list = new ArrayList<>(n);
        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            list.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            list.get(edge[0]).add(edge[1]);
            list.get(edge[1]).add(edge[0]);
        }
        return Math.max((dfs(list, 0, hasApple) - 1) * 2, 0);
    }
    
    public int dfs(List<List<Integer>> list, int node, List<Boolean> hasApple) {
        if (visited[node]) {
            return 0;
        }
        visited[node] = true;
        int cost = 0;
        for (Integer neigh : list.get(node)) {
            cost += dfs(list, neigh, hasApple);
        }
        if (hasApple.get(node) == false && cost == 0) {
            return 0;
        }
        return cost + 1;
    }
}